File: app/src/main/java/com/veganbeauty/app/features/community/message/MessageAdapter.java
package com.veganbeauty.app.features.community.message;

import android.graphics.Color;
import android.graphics.Typeface;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.veganbeauty.app.R;
import com.veganbeauty.app.data.local.entities.ConversationEntity;
import com.veganbeauty.app.data.local.entities.MemberInfoEntity;
import com.veganbeauty.app.utils.MessageTimeFormatter;
import com.veganbeauty.app.utils.RootieBrandHelper;

import java.util.ArrayList;
import java.util.List;

import coil.Coil;
import coil.request.ImageRequest;
import coil.transform.CircleCropTransformation;

public class MessageAdapter extends RecyclerView.Adapter<MessageAdapter.ViewHolder> {

    public interface OnMessageClickListener {
        void onClick(ConversationEntity conversation);
    }

    private final List<ConversationEntity> list;
    private final String currentUserId;
    private final OnMessageClickListener listener;

    public MessageAdapter(List<ConversationEntity> list, String currentUserId, OnMessageClickListener listener) {
        this.list = new ArrayList<>(list);
        this.currentUserId = currentUserId;
        this.listener = listener;
    }

    public void updateList(List<ConversationEntity> newList) {
        list.clear();
        if (newList != null) {
            list.addAll(newList);
        }
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.com_item_message, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        ConversationEntity item = list.get(position);
        
        String partnerId = "";
        if (item.getMembers() != null) {
            for (String memberId : item.getMembers()) {
                if (!memberId.equals(currentUserId)) {
                    partnerId = memberId;
                    break;
                }
            }
        }

        MemberInfoEntity partnerInfo = null;
        if (item.getMemberInfo() != null) {
            partnerInfo = item.getMemberInfo().get(partnerId);
        }

        if (holder.tvName != null) {
            holder.tvName.setText(partnerInfo != null ? partnerInfo.getName() : "Người dùng");
        }

        if (partnerInfo != null && holder.ivAvatar != null) {
            String avatarUrl = RootieBrandHelper.resolveAvatar(partnerId, partnerInfo.getAvatar());
            ImageRequest request = new ImageRequest.Builder(holder.itemView.getContext())
                    .data(avatarUrl)
                    .crossfade(true)
                    .transformations(new CircleCropTransformation())
                    .placeholder(R.drawable.img_avatar)
                    .error(R.drawable.img_avatar)
                    .target(holder.ivAvatar)
                    .build();
            Coil.imageLoader(holder.itemView.getContext()).enqueue(request);
        } else if (holder.ivAvatar != null) {
            holder.ivAvatar.setImageResource(R.drawable.img_avatar);
        }

        if (holder.tvLastMessage != null) {
            holder.tvLastMessage.setText(item.getLastMessage() != null ? item.getLastMessage() : "");
        }
        if (holder.tvTime != null) {
            String timeRaw = item.getLastMessageAt();
            if (timeRaw == null || timeRaw.trim().isEmpty()) {
                timeRaw = item.getUpdatedAt();
            }
            holder.tvTime.setText(MessageTimeFormatter.formatConversationTime(timeRaw));
        }

        boolean isUnread = item.getUnreadBy() != null && item.getUnreadBy().contains(currentUserId);
        if (holder.tvLastMessage != null) {
            if (isUnread) {
                holder.tvLastMessage.setTypeface(null, Typeface.BOLD);
                holder.tvLastMessage.setTextColor(Color.BLACK);
            } else {
                holder.tvLastMessage.setTypeface(null, Typeface.NORMAL);
                holder.tvLastMessage.setTextColor(Color.GRAY);
            }
        }
        if (holder.viewUnread != null) {
            holder.viewUnread.setVisibility(isUnread ? View.VISIBLE : View.GONE);
        }

        holder.itemView.setOnClickListener(v -> listener.onClick(item));
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        ImageView ivAvatar;
        TextView tvName;
        TextView tvLastMessage;
        TextView tvTime;
        View viewUnread;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            ivAvatar = itemView.findViewById(R.id.ivAvatar);
            tvName = itemView.findViewById(R.id.tvName);
            tvLastMessage = itemView.findViewById(R.id.tvLastMessage);
            tvTime = itemView.findViewById(R.id.tvTime);
            viewUnread = itemView.findViewById(R.id.vUnreadDot);
        }
    }
}

File: app/src/main/java/com/veganbeauty/app/features/community/message/MessageAdapter.kt
package com.veganbeauty.app.features.community.message

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import coil.load
import coil.transform.CircleCropTransformation
import com.veganbeauty.app.R
import com.veganbeauty.app.data.local.entities.ConversationEntity

class MessageAdapter(
    private var items: List<ConversationEntity>,
    private val onItemClick: (ConversationEntity) -> Unit
) : RecyclerView.Adapter<MessageAdapter.MessageViewHolder>() {

    class MessageViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val ivAvatar: ImageView = view.findViewById(R.id.ivAvatar)
        val vActiveDot: View = view.findViewById(R.id.vActiveDot)
        val tvName: TextView = view.findViewById(R.id.tvName)
        val tvLastMessage: TextView = view.findViewById(R.id.tvLastMessage)
        val tvTime: TextView = view.findViewById(R.id.tvTime)
        val vUnreadDot: View = view.findViewById(R.id.vUnreadDot)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MessageViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.com_item_message, parent, false)
        return MessageViewHolder(view)
    }

    private fun formatTimestamp(timestampStr: String): String {
        if (timestampStr.isEmpty()) return ""
        try {
            val format = java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", java.util.Locale.getDefault())
            format.timeZone = java.util.TimeZone.getTimeZone("UTC")
            val date = format.parse(timestampStr) ?: return ""

            val calendar = java.util.Calendar.getInstance()
            val now = java.util.Calendar.getInstance()
            calendar.time = date

            val formatTime = java.text.SimpleDateFormat("HH:mm", java.util.Locale("vi", "VN"))
            val formatDate = java.text.SimpleDateFormat("dd/MM", java.util.Locale("vi", "VN"))

            val timeStr = formatTime.format(calendar.time)

            return if (calendar.get(java.util.Calendar.YEAR) == now.get(java.util.Calendar.YEAR) &&
                calendar.get(java.util.Calendar.DAY_OF_YEAR) == now.get(java.util.Calendar.DAY_OF_YEAR)
            ) {
                timeStr
            } else {
                val dateStr = formatDate.format(calendar.time)
                "$timeStr $dateStr"
            }
        } catch (e: Exception) {
            return ""
        }
    }

    override fun onBindViewHolder(holder: MessageViewHolder, position: Int) {
        val item = items[position]
        
        val currentUserId = com.veganbeauty.app.data.local.ProfileSession.getCurrentUserId(holder.itemView.context)
        val partnerId = (item.members ?: emptyList()).firstOrNull { it != currentUserId } ?: ""
        val partnerInfo = (item.memberInfo ?: emptyMap())[partnerId]
        
        var partnerName = partnerInfo?.name ?: "Unknown"
        var partnerAvatar = partnerInfo?.avatar ?: ""
        
        // Fix for legacy 'You' data stored in db
        if (partnerName == "You" || partnerName == "Unknown" || partnerAvatar.isEmpty()) {
            val user = com.veganbeauty.app.data.local.LocalJsonReader(holder.itemView.context).getUsers().find { it.user_id == partnerId }
            if (user != null) {
                if (partnerName == "You" || partnerName == "Unknown") partnerName = user.full_name ?: partnerName
                if (partnerAvatar.isEmpty()) partnerAvatar = user.avatar ?: ""
            } else if (partnerId == "test_001") {
                if (partnerName == "You" || partnerName == "Unknown") partnerName = "Test User"
            }
        }
        
        val lastMsgText = item.lastMessage ?: ""
        val lastTime = formatTimestamp(item.lastMessageAt ?: "")
        
        holder.tvName.text = partnerName
        holder.tvLastMessage.text = lastMsgText
        holder.tvTime.text = lastTime

        if (partnerAvatar.isNotEmpty()) {
            holder.ivAvatar.load(partnerAvatar) {
                crossfade(true)
                placeholder(R.color.gray_light)
                if (partnerId == "rootie_vn") {
                    error(R.drawable.ic_logo_rootie)
                } else {
                    error(R.drawable.mascot_message)
                }
                transformations(CircleCropTransformation())
            }
        } else {
            if (partnerId == "rootie_vn") {
                holder.ivAvatar.setImageResource(R.drawable.ic_logo_rootie)
            } else {
                holder.ivAvatar.setImageResource(R.drawable.mascot_message)
            }
        }

        holder.vActiveDot.visibility = if ((item.activeBy ?: emptyList()).contains(partnerId)) View.VISIBLE else View.GONE
        
        val isUnread = (item.unreadBy ?: emptyList()).contains(currentUserId)
        if (isUnread) {
            holder.vUnreadDot.visibility = View.VISIBLE
            holder.tvName.setTypeface(null, android.graphics.Typeface.BOLD)
            holder.tvLastMessage.setTypeface(null, android.graphics.Typeface.BOLD)
            holder.tvLastMessage.setTextColor(holder.itemView.context.resources.getColor(R.color.black, null))
        } else {
            holder.vUnreadDot.visibility = View.GONE
            holder.tvName.setTypeface(null, android.graphics.Typeface.NORMAL)
            holder.tvLastMessage.setTypeface(null, android.graphics.Typeface.NORMAL)
            holder.tvLastMessage.setTextColor(holder.itemView.context.resources.getColor(R.color.gray_dark, null))
        }
        
        holder.itemView.setOnClickListener {
            onItemClick(item)
        }
    }

    override fun getItemCount() = items.size

    fun updateData(newItems: List<ConversationEntity>) {
        items = newItems
        notifyDataSetChanged()
    }
}
